File: onetrans/serving/pipeline.py
"""Nearline/Online 编排（两阶段服务 worker）。

- :class:`NearlineWorker`：Stage I，编码用户历史并写入 User KV（prefill）。
- :class:`OnlineWorker`：Stage II，读 User KV 并对候选交叉打分（decode）。
- :class:`BatchScheduler`：FIFO 攒批调度器，超时/满批触发批量打分（吞吐优先）。

worker 只依赖 :class:`TwoStageRunner`、:class:`KVStore` 与 :class:`ServingMetrics`，
与底层存储后端解耦（见 ``kv_store.py`` 的存储无关契约）。

增量 append 优化（复用旧 K/V 仅算新增位置）为后续里程碑；当前 nearline 采用
「全量 prefill + put」的正确基线（对应设计文档 §3.1 的幂等重建路径）。
"""

from __future__ import annotations

import threading
import time
from collections import deque
from dataclasses import dataclass
from typing import Any

import torch
from torch import Tensor

from onetrans.serving.kv_store import KVKey, KVStore, PutResult, UserKVRecord
from onetrans.serving.metrics import ServingMetrics
from onetrans.serving.serialize import deserialize, serialize
from onetrans.serving.two_stage import TwoStageRunner


class NearlineWorker:
    def __init__(
        self,
        runner: TwoStageRunner,
        store: KVStore,
        metrics: ServingMetrics | None = None,
        dtype: str = "float16",
    ) -> None:
        self.runner = runner
        self.store = store
        self.metrics = metrics or ServingMetrics()
        self.dtype = dtype

    def ingest(
        self,
        s_emb: Tensor,
        s_mask: Tensor,
        user_id: str,
        model_version: str,
        seq_ts_last: int = 0,
    ) -> PutResult:
        """编码用户历史并写入 User KV。

        :param s_emb: tokenize+pos+RMSNorm 后的 S 序列 [1, S0, D]
        :param s_mask: [1, S0] bool 有效掩码
        """
        with self.metrics.timing("nearline.encode_stage1"):
            kv = self.runner.encode_s(s_emb, s_mask)
        self.metrics.count("nearline.events_ingested", 1.0)

        with self.metrics.timing("nearline.append_kv"):
            payload = serialize(kv.per_layer, s_len=kv.s_len, per_layer_len=kv.per_layer_len)
            rec = UserKVRecord(
                key=KVKey(model_version=model_version, user_id=user_id),
                s_len=kv.s_len,
                per_layer_len=kv.per_layer_len,
                dtype=self.dtype,
                payload=payload,
                seq_ts_last=seq_ts_last,
                created_at=int(time.time()),
            )
            res = self.store.put(rec)
        if hasattr(self.store, "size"):
            self.metrics.gauge("kv.obj_cnt", float(self.store.size()))
        return res


class OnlineWorker:
    def __init__(
        self,
        runner: TwoStageRunner,
        store: KVStore,
        metrics: ServingMetrics | None = None,
    ) -> None:
        self.runner = runner
        self.store = store
        self.metrics = metrics or ServingMetrics()
        self._num_tasks = self.runner.linear.out_features

    def score(
        self,
        user_id: str,
        model_version: str,
        ns_emb: Tensor,
    ) -> Tensor:
        """读 User KV 并对候选 NS token 打分。

        :param ns_emb: tokenize+RMSNorm 后的 NS 序列 [B=M, Ns, D]
        :return: logits [M, T]
        """
        key = KVKey(model_version=model_version, user_id=user_id)
        with self.metrics.timing("online.kv_get"):
            rec = self.store.get(key)
        if rec is None:
            # KV miss（冷启动/过期/搬迁）：返回全零 legal logits，不抛异常（见 gap_analysis G8）
            self.metrics.count("kv.miss", 1.0)
            return torch.zeros(ns_emb.shape[0], self._num_tasks, device=ns_emb.device)
        self.metrics.count("kv.hit", 1.0)

        with self.metrics.timing("online.encode_stage2"):
            kv = decode_record(rec, ns_emb.device)
            logits = self.runner.score_ns(kv, ns_emb)
        self.metrics.count("online.qps", 1.0)
        self.metrics.count("online.candidate_throughput", float(ns_emb.shape[0]))
        return logits

    def score_batch(self, batch: list[ScoreRequest]) -> Tensor:
        """批量打分：一次 ``mget`` + 一次 ``score_ns_batch`` 打包多个请求。

        :param batch: 若干（user, 候选）打分请求（:class:`ScoreRequest`）
        :return: logits ``[B, T]``（B 为所有候选总数，顺序与 batch 展平一致）
        """
        keys = [r.key for r in batch]
        with self.metrics.timing("online.kv_mget"):
            recs = self.store.mget(keys)
        kvs: list[Any] = []
        embs: list[Tensor] = []
        miss_positions: list[int] = []
        flat = 0
        for rec, req in zip(recs, batch):
            m = req.ns_emb.shape[0]
            if rec is None:
                self.metrics.count("kv.miss", 1.0)
                miss_positions.extend(range(flat, flat + m))  # 记录 miss 候选的展平位置
            else:
                self.metrics.count("kv.hit", 1.0)
                kv = decode_record(rec, req.ns_emb.device)
                # 同用户多候选：kv 重复、候选展平
                for _ in range(m):
                    kvs.append(kv)
                embs.append(req.ns_emb)
            flat += m
        total = flat

        # miss 降级：命中候选正常打分，miss 候选填全零，保持 [B, T] 与展平顺序（见 gap_analysis G8）
        out = torch.zeros(total, self._num_tasks, device=batch[0].ns_emb.device)
        if kvs:
            ns_emb = torch.cat(embs, dim=0)  # [num_hits, Ns, D]
            with self.metrics.timing("online.encode_stage2"):
                logits = self.runner.score_ns_batch(kvs, ns_emb)  # [num_hits, T]
            miss_set = set(miss_positions)
            hit_ptr = 0
            for i in range(total):
                if i not in miss_set:
                    out[i] = logits[hit_ptr]
                    hit_ptr += 1
        self.metrics.count("online.qps", float(len(batch)))
        self.metrics.count("online.candidate_throughput", float(total))
        self.metrics.gauge("online.batch_size", float(total))
        return out


@dataclass
class ScoreRequest:
    """一次打分请求（一个 user 的 M 个候选）。"""

    key: KVKey
    ns_emb: Tensor  # [M, Ns, D]


class BatchScheduler:
    """FIFO 攒批调度器：满批或超时即吐出一个批次（动态 batching / 吞吐优先）。

    线程安全：生产者 :meth:`submit` 入队，消费者 :meth:`next_batch` 阻塞攒批。
    攒批窗口 ``max_wait_seconds`` 内若未满批，也按「已攒到的 ≥1 条」返回，
    保证时延有界。
    """

    def __init__(self, max_batch_size: int = 64, max_wait_seconds: float = 0.005) -> None:
        if max_batch_size <= 0:
            raise ValueError("max_batch_size 必须 ≥ 1")
        self.max_batch_size = max_batch_size
        self.max_wait_seconds = max_wait_seconds
        self._queue: deque[ScoreRequest] = deque()
        self._cond = threading.Condition()

    def submit(self, req: ScoreRequest) -> None:
        with self._cond:
            self._queue.append(req)
            self._cond.notify()

    def next_batch(self) -> list[ScoreRequest]:
        """阻塞取出一个批次（≥1 条，满批或超时返回）。"""
        with self._cond:
            while not self._queue:
                self._cond.wait()
            deadline = time.monotonic() + self.max_wait_seconds
            while len(self._queue) < self.max_batch_size:
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    break
                self._cond.wait(timeout=remaining)
            batch: list[ScoreRequest] = []
            while self._queue and len(batch) < self.max_batch_size:
                batch.append(self._queue.popleft())
            return batch

    def __len__(self) -> int:
        with self._cond:
            return len(self._queue)


def decode_record(rec: UserKVRecord, device: torch.device) -> Any:
    """把 UserKVRecord 反序列化为 :class:`UserKV`（供二阶段引擎直接消费）。"""
    from onetrans.serving.two_stage import UserKV

    per_layer = [(k.to(device), v.to(device)) for k, v in deserialize(rec.payload)]
    return UserKV(per_layer=per_layer, per_layer_len=rec.per_layer_len, s_len=rec.s_len)

File: cpp/src/serving/pipeline.cpp
#include "serving/pipeline.h"

#include <sstream>
#include <stdexcept>

namespace onetrans {

namespace {
int64_t now_unix() { return static_cast<int64_t>(std::time(nullptr)); }
}

// --------------------------------------------------------------------------- //
// Metrics
// --------------------------------------------------------------------------- //
void Metrics::count(const std::string& name, double v) {
    std::lock_guard<std::mutex> lk(mu_);
    values_[name] += v;
}

void Metrics::gauge(const std::string& name, double v) {
    std::lock_guard<std::mutex> lk(mu_);
    values_[name] = v;
}

double Metrics::get(const std::string& name) const {
    std::lock_guard<std::mutex> lk(mu_);
    auto it = values_.find(name);
    return it == values_.end() ? 0.0 : it->second;
}

Metrics::Timer::Timer(Metrics& m, std::string name) : m_(m), name_(std::move(name)) {
    t0_ = std::chrono::steady_clock::now();
}

Metrics::Timer::~Timer() {
    auto us = std::chrono::duration_cast<std::chrono::microseconds>(
                  std::chrono::steady_clock::now() - t0_)
                  .count();
    m_.count(name_ + "_us_total", static_cast<double>(us));
    m_.count(name_ + "_n", 1.0);
}

std::string Metrics::dump() const {
    std::unordered_map<std::string, double> snap;
    {
        std::lock_guard<std::mutex> lk(mu_);
        snap = values_;
    }
    std::ostringstream os;
    for (const auto& [k, v] : snap) os << k << " " << v << "\n";
    return os.str();
}

// --------------------------------------------------------------------------- //
// NearlineWorker
// --------------------------------------------------------------------------- //
IngestResult NearlineWorker::ingest(const IngestInput& in) {
    IngestResult res;
    try {
        Metrics::Timer timer(metrics_, "nearline.ingest");
        auto [s_emb, s_mask] = frontend_.encode_s(in.item_ids, lookup_);
        UserKV kv = runner_.encode_s(s_emb, s_mask);
        std::string payload = kv_serialize(kv);

        UserKVRecord rec;
        rec.key = KVKey{model_version_, in.user_id};
        rec.s_len = kv.s_len;
        rec.per_layer_len = kv.per_layer_len;
        rec.seq_ts_last = in.timestamps.empty() ? 0 : in.timestamps.back();
        rec.created_at = now_unix();
        rec.payload = std::move(payload);
        res.checksum = rec.checksum();
        res.accepted = store_.put(rec);

        metrics_.count("nearline.events_ingested", static_cast<double>(in.item_ids.size()));
        metrics_.gauge("kv.obj_cnt", static_cast<double>(store_.size()));
    } catch (const std::exception& e) {
        res.accepted = false;
        res.reason = e.what();
    }
    return res;
}

// --------------------------------------------------------------------------- //
// OnlineWorker
// --------------------------------------------------------------------------- //
ScoreOutcome OnlineWorker::score(const ScoreInput& in) {
    Metrics::Timer timer(metrics_, "online.score");
    Tensor ns_emb = frontend_.encode_ns(in, lookup_);
    const int64_t M = ns_emb.shape[0];
    const int64_t T = runner_.model().head_w.shape[0];

    auto rec = store_.get(KVKey{model_version_, in.user_id});
    if (!rec) {
        // KV miss（冷启动/过期/搬迁）：全零 legal logits 降级（对齐 G8 语义）
        metrics_.count("kv.miss");
        ScoreOutcome out;
        out.logits = Tensor::zeros({M, T});
        out.kv_hit = false;
        return out;
    }
    metrics_.count("kv.hit");

    UserKV kv = kv_deserialize(rec->payload);
    ScoreOutcome out;
    out.logits = runner_.score_ns(kv, ns_emb);
    metrics_.count("online.qps");
    metrics_.count("online.candidate_throughput", static_cast<double>(M));
    return out;
}

std::vector<ScoreOutcome> OnlineWorker::score_batch(const std::vector<ScoreInput>& batch) {
    Metrics::Timer timer(metrics_, "online.score_batch");
    if (batch.empty()) return {};

    // 1. 逐请求前处理
    std::vector<Tensor> ns_embs;
    ns_embs.reserve(batch.size());
    for (const auto& in : batch) ns_embs.push_back(frontend_.encode_ns(in, lookup_));

    // 2. 一次 mget
    std::vector<KVKey> keys;
    keys.reserve(batch.size());
    for (const auto& in : batch) keys.push_back(KVKey{model_version_, in.user_id});
    auto recs = store_.mget(keys);

    // 3. 命中请求：K/V 反序列化 + 候选展平；miss 请求记录位置
    std::vector<std::unique_ptr<UserKV>> kvs;
    std::vector<Tensor> hit_embs;
    std::vector<int> hit_of_req(batch.size(), -1);  // 请求 → 命中展平行
    int64_t total = 0;
    for (size_t r = 0; r < batch.size(); ++r) {
        const int64_t M = ns_embs[r].shape[0];
        if (recs[r]) {
            metrics_.count("kv.hit");
            kvs.push_back(std::make_unique<UserKV>(kv_deserialize(recs[r]->payload)));
            hit_embs.push_back(ns_embs[r]);
            hit_of_req[r] = static_cast<int>(hit_embs.size()) - 1;
        } else {
            metrics_.count("kv.miss");
        }
        total += M;
    }

    // 4. 一次 score_ns_batch：每候选一行（同用户重复指向其 KV）
    const int64_t T = runner_.model().head_w.shape[0];
    Tensor hit_logits;
    if (!kvs.empty()) {
        int64_t rows = 0;
        for (const auto& e : hit_embs) rows += e.shape[0];
        Tensor ns_cat({rows, ns_embs[0].shape[1], ns_embs[0].shape[2]});
        std::vector<const UserKV*> kv_rows;
        kv_rows.reserve(static_cast<size_t>(rows));
        int64_t off = 0;
        for (size_t h = 0; h < hit_embs.size(); ++h) {
            const auto& e = hit_embs[h];
            std::copy(e.data.begin(), e.data.end(), ns_cat.data.begin() + off);
            off += e.data.size();
            for (int64_t m = 0; m < e.shape[0]; ++m) kv_rows.push_back(kvs[h].get());
        }
        hit_logits = runner_.score_ns_batch(kv_rows, ns_cat);
    }

    // 5. 回填：miss 全零行，顺序与 batch 一致
    std::vector<ScoreOutcome> out(batch.size());
    for (size_t r = 0; r < batch.size(); ++r) {
        const int64_t M = ns_embs[r].shape[0];
        out[r].kv_hit = hit_of_req[r] >= 0;
        out[r].logits = Tensor::zeros({M, T});
        if (hit_of_req[r] >= 0) {
            int64_t row0 = 0;
            for (int h = 0; h < hit_of_req[r]; ++h) row0 += hit_embs[h].shape[0];
            std::copy(hit_logits.data.begin() + row0 * T, hit_logits.data.begin() + (row0 + M) * T,
                      out[r].logits.data.begin());
        }
    }
    metrics_.count("online.qps", static_cast<double>(batch.size()));
    metrics_.count("online.candidate_throughput", static_cast<double>(total));
    metrics_.gauge("online.batch_size", static_cast<double>(total));
    return out;
}

// --------------------------------------------------------------------------- //
// BatchScheduler
// --------------------------------------------------------------------------- //
void BatchScheduler::submit(ScoreJob job) {
    {
        std::lock_guard<std::mutex> lk(mu_);
        queue_.push_back(std::move(job));
    }
    cv_.notify_one();
}

std::vector<ScoreJob> BatchScheduler::next_batch() {
    std::unique_lock<std::mutex> lk(mu_);
    cv_.wait(lk, [&] { return !queue_.empty() || shutdown_; });
    if (queue_.empty()) return {};  // shutdown 唤醒
    auto deadline = std::chrono::steady_clock::now() + max_wait_;
    while (queue_.size() < max_batch_) {
        if (cv_.wait_until(lk, deadline) == std::cv_status::timeout) break;
    }
    std::vector<ScoreJob> batch;
    batch.reserve(std::min(max_batch_, queue_.size()));
    while (!queue_.empty() && batch.size() < max_batch_) {
        batch.push_back(std::move(queue_.front()));
        queue_.pop_front();
    }
    return batch;
}

void BatchScheduler::shutdown() {
    {
        std::lock_guard<std::mutex> lk(mu_);
        shutdown_ = true;
    }
    cv_.notify_all();
}

size_t BatchScheduler::pending() const {
    std::lock_guard<std::mutex> lk(mu_);
    return queue_.size();
}

// --------------------------------------------------------------------------- //
// Dispatcher
// --------------------------------------------------------------------------- //
Dispatcher::Dispatcher(OnlineWorker& worker, BatchScheduler& sched, int num_threads,
                       Metrics& metrics)
    : worker_(worker), sched_(sched), metrics_(metrics), num_threads_(num_threads) {
    if (num_threads_ <= 0) num_threads_ = 1;
}

Dispatcher::~Dispatcher() { stop(); }

std::future<ScoreOutcome> Dispatcher::submit(ScoreInput input) {
    auto promise = std::make_shared<std::promise<ScoreOutcome>>();
    auto fut = promise->get_future();
    sched_.submit(ScoreJob{std::move(input), promise});
    return fut;
}

void Dispatcher::start() {
    if (running_.exchange(true)) return;
    for (int i = 0; i < num_threads_; ++i) threads_.emplace_back([this] { worker_loop(); });
}

void Dispatcher::stop() {
    if (!running_.exchange(false)) return;
    sched_.shutdown();
    for (auto& t : threads_) {
        if (t.joinable()) t.join();
    }
    threads_.clear();
}

void Dispatcher::worker_loop() {
    while (running_) {
        auto batch = sched_.next_batch();
        if (batch.empty()) break;  // shutdown
        std::vector<ScoreInput> inputs;
        inputs.reserve(batch.size());
        for (auto& job : batch) inputs.push_back(std::move(job.input));
        try {
            auto outcomes = worker_.score_batch(inputs);
            for (size_t i = 0; i < batch.size(); ++i)
                batch[i].promise->set_value(std::move(outcomes[i]));
        } catch (const std::exception& e) {
            for (auto& job : batch)
                job.promise->set_exception(std::make_exception_ptr(std::runtime_error(e.what())));
        }
    }
}

}  // namespace onetrans
